jerry-core/ecma/operations/ecma-promise-object.c

```diff
@@ -0,0 +1,707 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ecma-boolean-object.h"
+#include "ecma-builtins.h"
+#include "ecma-exceptions.h"
+#include "ecma-function-object.h"
+#include "ecma-gc.h"
+#include "ecma-globals.h"
+#include "ecma-helpers.h"
+#include "ecma-jobqueue.h"
+#include "ecma-objects.h"
+#include "ecma-objects-general.h"
+#include "ecma-promise-object.h"
+#include "ecma-try-catch-macro.h"
+#include "jmem.h"
+
+#ifndef CONFIG_DISABLE_ES2015_PROMISE_BUILTIN
+
+/** \addtogroup ecma ECMA
+ * @{
+ *
+ * \addtogroup ecmapromiseobject ECMA Promise object related routines
+ * @{
+ */
+
+/**
+ * Check if a object is promise
+ *
+ * @return bool
+ */
+inline bool
+ecma_is_promise (ecma_object_t *obj_p) /**< points to object */
+{
+  lit_magic_string_id_t class_id = ecma_object_get_class_name (obj_p);
+
+  return class_id == LIT_MAGIC_STRING_PROMISE_UL;
+} /* ecma_is_promise */
+
+/**
+ * Get the result of the promise
+ *
+ * @return ecma value of the result
+ */
+inline ecma_value_t
+ecma_promise_get_result (ecma_object_t *obj_p) /**< points to promise object */
+{
+  ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) obj_p;
+
+  return ecma_copy_value (ext_object_p->u.class_prop.u.value);
+} /* ecma_promise_get_result */
+
+/**
+ * Set the PromiseResult of promise
+ */
+inline void
+ecma_promise_set_result (ecma_object_t *obj_p, /**< points to promise object */
+                         ecma_value_t result) /**< the result value */
+{
+  ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) obj_p;
+
+  ext_object_p->u.class_prop.u.value = ecma_copy_value_if_not_object (result);
```

Comment:
I plan to remove the `ecma_copy_value_if_not_object` in the body. caller should do this before call the ecma_promise_set_result. It is similar with ecma_op_object_put.